Implement IOGroup fallback in PlatformIO

bgeltz opened the pull request:
- Relates to #2130, #1690.
- Fixes #1813.

- This affects the following APIs: read_signal(), write_control(), push_signal(), push_control().
- When a signal or control is requested through one of the above APIs, all IOGroups will be evaluated to see if they support the request.
  + The domain of the last registered plugin in the IOGroupFactory that supports the control or signal is used to determine the "native domain".  Any subsequent IOGroups that do not match this domain exactly are disregarded.
  + When read_signal() or write_control() is invoked, if there is an issue with the attempted call with the first IOGroup, the next in the list will be evaluated until all IOGroups that have been previously identified are exhausted.  An exception is raised in this instance.
  + Otherwise the IOGroup that is able to service the request will do so.

bgeltz committed:
Update PIO methods to step through IOGroups

- Introduce fallback logic if there is an issue with
  trying to use a particular IOGroup.
- Now all IOGroups that support a given signal or control
  at the same domain will be evaluated if there is an issue.
- If the domain changes between IOGroups, that IOGroup is not
  used.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Remove bad includes

- Replace with actually used includes.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Rearrange PIO includes

- They now follow the following ordering:
    config.h, PIOImp.hpp,
    C includes, STL C++ includes,
    GEOPM public includes,
    GEOPM private C includes, GEOPM private C++ includes.
- Alphabetically sorted for each sub group.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Refactor PIO::find_*_iogroup to return vectors

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -560,55 +632,55 @@ namespace geopm
-        auto iogroup = find_signal_iogroup(signal_name);
-        if (iogroup == nullptr) {
+        auto iogroups = find_signal_iogroup(signal_name);
+        if (iogroups.size() == 0) {
```
.empty()?

bgeltz committed:
Address review feedback

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Attempt to read_signal when pushing signals

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Attempt to read_signal/write_control when pushing controls

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Handle issues around reading controls as signals in IOGroups

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add iogroup fallback tests for PIO::push_signal

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add iogroup fallback tests for PIO::write_control

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Address review feedback

- Minimize the try/catch block in the push_* APIs.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Handle IOGroups that do not support read_signal()/write_control()

- Use GEOPM_ERROR_NOT_IMPLEMENTED to signify that
  these methods are intentionally not implemented
  and PIO should allow the push_* to proceed.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Warn when the native domain changes between IOGroups

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -105,31 +102,61 @@ namespace geopm
+                }
+#ifdef GEOPM_DEBUG
+                else {
+                    std::cerr << "Warning: <geopm> PlatformIO::find_signal_iogroup(): "
```
Suggested warning:

Native domain differs for the same signal provided by another IOGroup.
The current implementation does not support fallback using an IOGroup
that provides a signal with a different native domain.  If this is an issue
for your use case, please request this feature.  The signal << signal_name <<
will not use << (*it)->name() for fallback if a read fails.

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -105,31 +102,61 @@ namespace geopm
+                }
+#ifdef GEOPM_DEBUG
+                else {
+                    std::cerr << "Warning: <geopm> PlatformIO::find_control_iogroup(): "
```
Similar warning as above might be clearer

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -210,26 +227,49 @@ namespace geopm
+                    try {
+                        // Attempt to read before pushing to ensure batch reads will succeed
+                        (void)ii->read_signal(signal_name, domain_type, domain_idx);
+                        read_signal_ok = true;
```
suggested rename:

read_signal_ok -> do_push_signal

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -210,26 +227,49 @@ namespace geopm
             result = sig_tup_it->second;
         }
+        std::string err_msg;
+        bool read_signal_ok = false;
```
Move this declaration closer to use: inside of for and if contexts, i.e. after line 235

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -296,26 +336,52 @@ namespace geopm
             result = ctl_tup_it->second;
         }
+        std::string err_msg;
+        bool read_signal_ok = false;
```
read_signal_ok -> is_write_valid

also move declaration after line 344

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -210,26 +227,49 @@ namespace geopm
-                    result = m_active_signal.size();
-                    m_existing_signal[sig_tup] = result;
-                    m_active_signal.emplace_back(iogroup, group_signal_idx);
+            auto iogroups = find_signal_iogroup(signal_name);
+            for (auto ii : iogroups) {
```
```suggestion
            for (auto ii : find_signal_iogroup(signal_name)) {
```

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -296,26 +336,52 @@ namespace geopm
-                    result = m_active_control.size();
-                    m_existing_control[ctl_tup] = result;
-                    m_active_control.emplace_back(iogroup, group_control_idx);
+            auto iogroups = find_control_iogroup(control_name);
+            for (auto ii : iogroups) {
```
```suggestion
            for (auto ii : find_control_iogroup(control_name)) {
```

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -296,26 +336,52 @@ namespace geopm
+                    }
+                    catch (const geopm::Exception &ex) {
+                        if (ex.err_value() == GEOPM_ERROR_NOT_IMPLEMENTED) {
+                            // IOGroups may not support read_signal()/write_control()
```
I wonder if we would actually support an IOGroup with a control that could only be written through the batch interface.  We do not currently have any IOGroups which are implemented this way.

bgeltz replied:
Here I was just referencing the fact that you cannot ```write_control()``` to either the EpochIOGroup or the ProfileIOGroup.

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -296,26 +336,52 @@ namespace geopm
+                        result = m_active_control.size();
+                        m_existing_control[ctl_tup] = result;
+                        m_active_control.emplace_back(ii, group_control_idx);
+                        break;
```
can we get rid of the early breaks and just put a

```
if (result != -1) {
    break;
}
```
At the end of the two loops over IOGroups in push_signal() and push_control()

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -438,17 +504,38 @@ namespace geopm
         }
-        else {
-            result = iogroup->read_signal(signal_name, domain_type, domain_idx);
+        if (std::isnan(result)) {
```
We need a different variable to track if the read was successful.  NAN may be returned when no exception was raised.

cmcantalupo commented on service/src/PlatformIO.cpp:
```diff
@@ -489,16 +576,39 @@ namespace geopm
-            write_control_convert_domain(control_name, domain_type, domain_idx, setting);
+
+        std::string err_msg;
+        bool write_complete = false;
```
write_complete -> is_write_complete

cmcantalupo reviewed (commented):
Looks pretty good.  One real issue:

https://github.com/geopm/geopm/pull/2144/files#r826474099

The rest are style related.

bgeltz committed:
Address review feedback

- Update warning message.
- Refactor: rename and simplify.
- Don't use NAN to gate returns.
- Simplify usage of break.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Address review feedback

- Move frequency value caching back to save_control().
- Call save_control() in the constructor.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add fallback tests for PIO::push_control()

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
